src/main/java/net/neoforged/neoforge/registries/DeferredHolder.java

```diff
@@ -0,0 +1,270 @@
+     * @throws NullPointerException  If the underlying Holder has not been populated (the target object is not registered).
+     */
+    @Override
+    public T get() {
+        return this.value();
+    }
+
+    /**
+     * @return An optional containing the target object, if {@link #isPresent()}, otherwise {@linkplain Optional#empty() an empty optional}.
+     */
+    public Optional<T> asOptional() {
+        return isPresent() ? Optional.of(value()) : Optional.empty();
+    }
+
+    /**
+     * The type of the registry is really <? super T> but this saves us some additional ugly casting and doesn't break anything.
+     *
+     * @return The registry that this DeferredHolder is pointing at, or null if it doesn't exist.
+     */
+    @Nullable
+    @SuppressWarnings("unchecked")
+    protected Registry<R> getRegistry() {
+        return (Registry<R>) BuiltInRegistries.REGISTRY.get(this.key.registry());
+    }
+
+    /**
+     * Binds this DeferredHolder to the underlying registry and target object.<br> Has no effect if already bound.
+     *
+     * @param throwOnMissingRegistry If true, an exception will be thrown if the registry is absent.
+     * @throws IllegalStateException If throwOnMissingRegistry is true and the backing registry is unavailable.
+     */
+    protected void bind(boolean throwOnMissingRegistry) {
+        if (this.holder != null) return;
+
+        Registry<R> registry = getRegistry();
+        if (registry != null) {
+            this.holder = registry.getHolder(this.key).orElse(null);
+        } else if (throwOnMissingRegistry) {
+            throw new IllegalStateException("Registry not present for " + this + ": " + this.key.registry());
+        }
+    }
+
+    /**
+     * @return The ID of the object pointed to by this DeferredHolder.
+     */
+    public ResourceLocation getId() {
+        return this.key.location();
+    }
+
+    /**
+     * @return The ResourceKey of the object pointed to by this DeferredHolder.
+     */
+    public ResourceKey<R> getKey() {
+        return this.key;
+    }
+
+    /**
+     * Gets the inner holder. Only exists for encapsulation reasons.
+     *
+     * @return the inner holder.
+     */
+    @Nullable
+    protected final Holder<R> getHolder() {
+        return this.holder;
+    }
+
+    /**
+     * @return True if this DeferredHolder has been bound, and {@link #get()} or {@link #value()} may be called.
+     */
+    public boolean isPresent() {
+        bind(false);
+        return this.holder != null;
+    }
+
+    @Override
+    public boolean equals(Object obj) {
+        if (this == obj) return true;
+        return obj instanceof DeferredHolder<?, ?> dh && dh.key == this.key;
+    }
+
+    @Override
+    public int hashCode() {
+        return this.key.hashCode();
+    }
+
+    @Override
+    public String toString() {
+        return String.format(Locale.ENGLISH, "DeferredHolder{%s}", this.key);
```

Comment:
Why the locale? I've never seen that anywhere.

Reply:
No idea, I have it there in the original but I'm not sure why its there either

Reply:
IIRC, adding `Locale.ENGLISH` to all (Neo/Forge) calls to `String` methods which take in a `Locale` (`format`, `toLowerCase`, `toUpperCase`) was done in a pass a good while ago, to prevent possible localization issues (good ol' [Turkish `i` conversion](https://garygregory.wordpress.com/2015/11/03/java-lowercase-conversion-turkey/)).